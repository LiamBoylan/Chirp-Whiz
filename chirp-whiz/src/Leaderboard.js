import React, { useState, useEffect } from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import Table from 'react-bootstrap/Table';

export function totalPercent(user) {
    let percent = 0;
    let corrTotal = user.correct.reduce((a, b) => a + b, 0);
    let incorrTotal = user.incorrect.reduce((a, b) => a + b, 0);

    if (corrTotal > 0 || incorrTotal > 0) {
        percent = (corrTotal/(corrTotal+incorrTotal))*100;
    }

    return percent.toFixed(2);
}

export function scorify(users) {
    let newUserArray = [];
    for (let i = 0; i < users.length; i++) {
        newUserArray.push({
            name: users[i].name,
            score: totalPercent(users[i])
        })
    }
    return newUserArray;
}

function Leaderboard(props) {
    const [lboardUsers, setLboardUsers] = useState([{
        name: 'name',
        score: 0.00
    }]);

    useEffect(() => {
        let usersWithScores = scorify(props.users);
        usersWithScores.sort(function(a, b){return b.score - a.score});
        setLboardUsers(usersWithScores);
    }, [])

    return (
        <div>
            <Table striped bordered hover>
            <thead>
                <tr>
                    <th>Username</th>
                    <th>Accuracy</th>
                </tr>
            </thead>
            <tbody>
                {lboardUsers.map((currentUser, i) =>
                    <tr key={i}>
                        <th>{currentUser.name}</th>
                        <th>{currentUser.score}%</th>
                    </tr>
                )}
            </tbody>
            </Table>
        </div>
    );
}

export default Leaderboard;