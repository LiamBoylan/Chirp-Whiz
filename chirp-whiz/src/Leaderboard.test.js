import React from 'react';
import Leaderboard from './Leaderboard';
import '@testing-library/jest-dom/extend-expect'
import { render } from '@testing-library/react'

const LeaderboardTest = require('./Leaderboard.js');

// Unit tests

test('mock Number and Array function for totalPercent testing', () => {
    const spy = jest.spyOn(Number.prototype, 'toFixed')
    const spy2 = jest.spyOn(Array.prototype, 'reduce')
    spy.mockReturnValueOnce(3);
    spy2.mockReturnValueOnce(4);
  
    expect(LeaderboardTest.totalPercent({correct: [0,0,0,], incorrect: [0,0,0]})).toBe(3);
})

test('mock Array.push and totalPercent for scorify testing', () => {
    const spy = jest.spyOn(Array.prototype, 'push')
    const spy2 = jest.spyOn(LeaderboardTest, 'totalPercent')
    spy.mockReturnValueOnce(3);
    spy2.mockReturnValueOnce(4);
  
    expect(LeaderboardTest.scorify({correct: [0,0,0], incorrect: [0,0,0]})).toStrictEqual([]);
})

test('mock Array.push to test for its usage in functions', () => {
    const spy = jest.spyOn(Array.prototype, 'push')
  
    expect(spy).toHaveBeenCalledTimes(5);
    spy.mockRestore();
})

// Integration tests

it('Leaderboard snapshot', () => {
    const container = render(<Leaderboard
        users={{
            name: 'l',
            email: 'l',
            correct: [0,0,0],
            incorrect: [0,0,0]
        }}
    />)

    expect(container.firstChild).toMatchSnapshot();
});

test("check if category in leaderboard renders", () => {
    const {getByText} = render(<Leaderboard
        users={{
            name: 'l',
            score: 0.00
        }}
    />)

    expect(getByText("Username")).toBeInTheDocument();
});

test("check if category in leaderboard renders", () => {
    const {getByText} = render(<Leaderboard
        users={{
            name: 'l',
            score: 0.00
        }}
    />)

    expect(getByText("Username")).toBeInTheDocument();
});