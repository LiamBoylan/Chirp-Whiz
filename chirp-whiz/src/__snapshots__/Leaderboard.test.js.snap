// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard snapshot 1`] = `undefined`;
